chalk-ir/src/visit.rs

```diff
@@ -0,0 +1,421 @@
+//! Traits for visiting bits of IR.
+use std::fmt::Debug;
+
+use crate::{
+    BoundVar, DebruijnIndex, DomainGoal, Goal, InferenceVar, Interner, Lifetime, LifetimeData,
+    PlaceholderIndex, ProgramClause, Ty, TyData, WhereClause,
+};
+
+mod binder_impls;
+mod boring_impls;
+pub mod visitors;
+
+pub use visitors::VisitExt;
+
+pub trait VisitResult: Sized {
+    fn new() -> Self;
+
+    fn return_early(&self) -> bool;
+    fn combine(self, other: Self) -> Self;
+
+    fn and_then(self, op: impl FnOnce() -> Self) -> Self {
+        if self.return_early() {
+            self
+        } else {
+            self.combine(op())
+        }
+    }
+}
+
+impl VisitResult for () {
```

Comment:
Comment:

Unit type for a visitor indicates a "side-effecting" visitor that should visit an entire term.